### PR TITLE
Fix xcmp_fee for automation-time

### PR DESCRIPTION
**Pull request by imstar15**

### CODE REVIEW

**Review by chrisli30 — CHANGES_REQUESTED** [review]
@imstar15 can you add comments on this webpage pointing out the core changes of this PR?

1. What is the TransactInfo before change? What are the keys and values?
2. What is the TransactInfo after? What are the keys and values?
3. If any extrinsic interface has changed, please provide screenshots for before and after.

Adding @v9n for a look.

**Comment by chrisli30** [comment]
In addition, "This branch is out-of-date with the base branch". Please rebase master.

**Comment by chrisli30** [comment]
We need to keep optimizing the Flow parameter in the transact_info for better functionality. Rather than specifying the flow for a destination, we should specify the supported flows instead. Most destination could support both flows.

In practice, this means that we will have separate extrinsics for each flow. The **scheduleXcmpTask** function would utilize the Normal flow, while **scheduleXcmpTaskThroughDerived** would utilize the Alternate flow. By distinguishing between the flows this way, we can achieve more efficient and targeted execution based on each task type’s requirements.

Furthermore, it's important to note that the current names, Normal and Alternate, lack clear meaning. The key distinction between them lies in who pays for the fees. In the Normal flow, Turing's sovereign account covers the XCM fee through ReserveAssetDeposited. On the other hand, in the Alternate flow, the user's proxy wallet bears the XCM fee. Another key difference is that the Alternate flow requires the user to pay XCM fees on the receiving end, thereby reducing the required amount of tokens on Turing (we only need to charge for task scheduling). However, whether users truly care about the slight fee difference remains uncertain.

You can find detailed instructions for each flow in the following code links:

**Normal flow instructions**
https://github.com/OAK-Foundation/OAK-blockchain/blob/67b4f7f5976d58dd35b080cf325eb01dff4921a7/pallets/xcmp-handler/src/lib.rs#L320

**Alternate flow instructions**
https://github.com/OAK-Foundation/OAK-blockchain/blob/67b4f7f5976d58dd35b080cf325eb01dff4921a7/pallets/xcmp-handler/src/lib.rs#L386C1-L386C1

While the name "Flow" is acceptable, it's worth considering more descriptive options since these sequences, combinations, or templates have distinct characteristics. I propose renaming the Flow selector to `XcmTaskSupported`, which consists of two types: `scheduleWithPrepaidFees ` for the normal flow and `scheduleWithPrepaidFees` for the alternate flow. Among the, the PrepaidFeeds refer to both the XCM fee and execution fee at the receiving end.

**Comment by chrisli30** [comment]
Discussed the new flow parametes with Charles, we don’t need the chain storage `XcmTaskSupported` as of now, so after this PR there should be no XCM task relevant parameter left in chain storage.

**Comment by imstar15** [comment]
1, What is the TransactInfo before change? What are the keys and values?
2. What is the TransactInfo after? What are the keys and values?

We have remove TransactInfo stroage.
User should input `overallWeight` and `executionFee` params for scheduleXcmpTask and scheduleXcmpTaskThroughProxy extrinsic.

```
overallWeight = encodedCallWeight + instructionWeight * instructionCount
executionFee =  FEE_PER_SECOND * overallWeight / WEIGHT_PER_SECOND
```

ScheduleXcmpTask run with scheduleWithPrepaidFees flow, the xcmp fee will be deducted in turing chain.
ScheduleXcmpTaskThroughProxy run with scheduleWithoutPrepaidFees flow, the xcmp fee will *not* be deducted in turing chain.

3. If any extrinsic interface has changed, please provide screenshots for before and after.

Changes:
1. paraId -> destination: Multilation
2. currencyId -> scheduleFee
3. xcmAssetLocation -> executionFee, AssetPayment { asset_location,  amount }
4.  overallWeight, new Parameter

Before:
image

After:
image

**Review by chrisli30 — APPROVED** [review]
The PR explanation looks clear to me. Good to merge.

File: pallets/automation-time/rpc/runtime-api/src/lib.rs [hunk 1]
```diff
@@ -45,8 +45,8 @@ pub enum AutomationAction {
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
 pub struct FeeDetails<Balance> {
+	pub schedule_fee: Balance,
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply pallets/automation-time/rpc/runtime-api/src/lib.rs:48]: Yes, we need to do two things about these changes.
1. Create a changelog list for existing dapp partners, and ask them to test the new extrinsics in Turing Staging.
2. Update docs.oak.tech with the latest details of extrinsics.

File: pallets/automation-time/src/types.rs [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
-use cumulus_primitives_core::ParaId;
-
 use pallet_automation_time_rpc_runtime_api::AutomationAction;
 
 use xcm::{latest::prelude::*, VersionedMultiLocation};
 
 pub type Seconds = u64;
 pub type UnixTime = u64;
 
+/// The struct that stores all information needed for a task.
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/types.rs:15]: Yes, It's a wrong comment.

File: pallets/automation-time/src/migrations/update_xcmp_task.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+				..
+			} => {
+				let schedule_fee =
+					T::CurrencyIdConvert::convert(currency_id).expect("IncoveribleCurrencyId");
+				Self::XCMP {
+					destination: MultiLocation::new(1, X1(Parachain(para_id.into()))),
+					schedule_fee,
+					execution_fee: AssetPayment {
+						asset_location: MultiLocation::new(0, Here).into(),
+						amount: 3000000000,
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/migrations/update_xcmp_task.rs:87]: The value of 3,000,000,000 is the execution fee for the Mangata XCMP task. Since only Mangata XCMP tasks exist on the chain currently, we only need to perform the migration for this type of task.

Here's how the value is calculated:

XCMP execution fee = FEE_PER_SECOND * (encodedCallWeight + xcmpInstructionWeight * xcmpInstructionCount) / WEIGHT_PER_SECOND

FEE_PER_SECOND: We use TUR as the payment currency for XCMP execution fee. In the Mangata-Kusama assetRegistry, the configured FEE_PER_SECOND for TUR is 537,600,000,000.

image

WEIGHT_PER_SECOND: In the Substrate framework, this is a fixed value of 1,000,000,000,000.

encodedCallWeight: This value can be obtained by running the script. It is derived from the object {refTime: 2905398711, proofSize: 3716}. For now, we only need the value of refTime.

```
const ACCOUNT = "5HTursw8Y2cevzj2PVDAkSF6Lv5snEtiU8fyNbpZFViMTf4b";
const compoundExtrinsic = api.tx.xyk.compoundRewards(5, 1000000)
const extrinsic = api.tx.proxy.proxy('5HTursw8Y2cevzj2PVDAkSF6Lv5snEtiU8fyNbpZFViMTf4b', 'AutoCompound', compoundExtrinsic);
console.log(await extrinsic.paymentInfo(ACCOUNT))
```

image

xcmpInstructionWeight: In Mangata, the instructionWeight is 150,000,000.

https://github.com/mangata-finance/mangata-node/blob/9595baea52f1ce2ed408301daa00dbfec504adfa/runtime/mangata-kusama/src/xcm_config.rs#L120

xcmpInstructionCount: We need 6 XCM instructions to execute this task.

Therefore:

XCMP execution fee = FEE_PER_SECOND * (encodedCallWeight + xcmpInstructionWeight * xcmpInstructionCount) / WEIGHT_PER_SECOND

XCMP execution fee = 537,600,000,000 * (2905398711 + 150,000,000 * 6) / 1,000,000,000,000 = 2,045,782,347

To ensure successful task execution, I set the fee to 3,000,000,000, slightly higher than the actual required fee.

Since the previous storage did not include the TUR's FEE_PER_SECOND, we cannot calculate it programmatically. 

This migration is only for the current upgrade, and these values will be used only once. Therefore, there is no need to put it into a constant.

File: pallets/automation-time/src/migrations/update_xcmp_task.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+					},
+					encoded_call,
+					encoded_call_weight: encoded_call_weight.clone(),
+					overall_weight: encoded_call_weight
+						.saturating_add(Weight::from_ref_time(1_000_000_000).saturating_mul(6)),
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/migrations/update_xcmp_task.rs:92]: Please see https://github.com/OAK-Foundation/OAK-blockchain/pull/366#discussion_r1253852787

File: pallets/xcmp-handler/src/tests.rs [hunk 5]
```diff
@@ -14,285 +14,57 @@
...
-	});
-}
 
 //*****************
 //Helper  functions
 //*****************
 
-// calculate_xcm_fee_and_weight
-#[test]
-fn calculate_xcm_fee_and_weight_works() {
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/xcmp-handler/src/tests.rs:130]: In this PR, I have removed the code for automatically calculating XCMP weight and execution fee. 

Users are now required to calculate the weight and fee themselves. Therefore, this test is no longer necessary.

File: pallets/automation-time/src/fees.rs [hunk 6]
```diff
@@ -73,45 +73,50 @@ where
...
+						T::CurrencyIdConvert::convert(self.schedule_fee_currency_id.into())
 							.ok_or("IncoveribleCurrencyId")?,
 					),
-					fun: Fungibility::Fungible(self.execution_fee.saturated_into()),
+					fun: Fungibility::Fungible(self.schedule_fee_amount.saturated_into()),
 				});
 
-				if self.xcmp_fee > MultiBalanceOf::<T>::zero() {
+				if self.execution_fee_amount > MultiBalanceOf::<T>::zero() {
 					T::XcmpTransactor::pay_xcm_fee(
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/fees.rs:116]: `pay_xcm_fee` is a suitable name for the function within the XcmpTransactor module.

File: node/src/chain_spec/turing.rs [hunk 7]
```diff
@@ -127,7 +91,16 @@ pub fn turing_development_config() -> ChainSpec {
 								name: b"Mangata Rococo".to_vec(),
 								symbol: b"MGR".to_vec(),
 								existential_deposit: Default::default(),
-								location: Some(MultiLocation::new(1, X1(Parachain(2110))).into()),
+								location: Some(
+									MultiLocation::new(
+										1,
+										X2(
+											Parachain(2110),
```
[the remark this answers is not in the document]
Comment (imstar15) [reply node/src/chain_spec/turing.rs:98]: The deleted lines are related to the removal of the data in xcmp_handler_asset_config. These lines modify the erroneous configuration of the Mangata asset in the assetRegistry module. The change is from `X1 (Parachain(2110))` to `X2 (Parachain(2110), GeneralKey { length: 4, data: [0; 32] })`.

File: pallets/automation-time/src/lib.rs [hunk 8]
```diff
@@ -1479,23 +1527,27 @@ pub mod pallet {
 		///
 		/// Fee saturates at Weight/BalanceOf when there are an unreasonable num of executions
 		/// In practice, executions is bounded by T::MaxExecutionTimes and unlikely to saturate
-		pub fn calculate_execution_fee(
+		pub fn calculate_schedule_fee_amount(
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply pallets/automation-time/src/lib.rs:1530]: Yes, this is a good call. We should test it in this PR.

File: pallets/automation-time/src/mock.rs [hunk 9]
```diff
@@ -50,6 +50,9 @@ pub const NATIVE: CurrencyId = 0;
 pub const NATIVE_LOCATION: MultiLocation = MultiLocation { parents: 0, interior: Here };
 pub const FOREIGN_CURRENCY_ID: CurrencyId = 1;
 
+pub const FEE_PER_SECOND: u128 = 100_000_000_000;
+pub const XCMP_TASK_REF_TIME: u128 = 20_000;
```
Comment (chrisli30) [inline pallets/automation-time/src/mock.rs:54]: You need to map the fee_per_second to destination, so we have 

dest1: {
    weight,
    refTime,
    assets: [ {token 1, fee per second 1}, {token 2, fee per second 2} ]
}

dest2 ..

and test them.

File: pallets/automation-time/src/tests.rs [hunk 10]
```diff
@@ -289,6 +293,33 @@ fn schedule_native_transfer_works() {
 	})
 }
 
+#[test]
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:296]: This happy path looks okay, and you need to create others to test its edge cases. For example,
1. A different schedule_fee asset should result in a different fee.
2. A different execution_fee asset should result in a different fee.
3.  A different schedule_as should result in a same fee.
4. A different flow should result in a different fee.

cc @v9n for a look.

File: pallets/automation-time/src/tests.rs [hunk 11]
```diff
@@ -289,6 +293,33 @@ fn schedule_native_transfer_works() {
...
+			encoded_call_weight: Weight::from_ref_time(100_000),
+			overall_weight: Weight::from_ref_time(200_000),
+			schedule_as: Some(delegator_account),
+			flow: InstructionSequence::PayThroughRemoteDerivativeAccount,
+		};
+		let schedule_fee_amount =
+			AutomationTime::calculate_schedule_fee_amount(&action, execution_count)
+				.expect("Calculate schedule fee amount should work");
+		let expected_schedule_fee_amount =
+			FEE_PER_SECOND * XCMP_TASK_REF_TIME * (execution_count as u128) /
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:317]: `action` is what you passed in to the function calculate_schedule_fee_amount. You need to use the variable values within **action** for calculation, instead of using a separate const variable XCMP_TASK_REF_TIME. XCMP_TASK_REF_TIME is not used in mock.

Comment (chrisli30) [reply pallets/automation-time/src/tests.rs:317]: WEIGHT_REF_TIME_PER_SECOND is a global variable for our chain, so that’s the only constant variable which can be used in this line of validation.

Fee per second is in relation to each destination, you need to mock and  to get them by destination.fee_per_second.

File: pallets/automation-time/src/tests.rs [hunk 12]
```diff
@@ -330,11 +362,15 @@ fn schedule_xcmp_through_proxy_works() {
...
-			MultiLocation::new(1, X1(Parachain(PARA_ID.into()))).into(),
+			Box::new(destination.clone().into()),
+			Box::new(MultiLocation::default().into()),
+			Box::new(AssetPayment {
+				asset_location: destination.into(),
+				amount: 10,
+			}),
 			call.clone(),
 			Weight::from_ref_time(100_000),
+			Weight::from_ref_time(200_000),
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:373]: Can you explain the reason of changing this value from 100_000 to 200_000?

Do we need two lines here?
```
Weight::from_ref_time(100_000),
Weight::from_ref_time(200_000),
```

File: pallets/automation-time/src/tests.rs [hunk 13]
```diff
@@ -374,11 +411,12 @@ fn schedule_xcmp_through_proxy_same_as_delegator_account() {
...
 				ScheduleParam::Fixed { execution_times: vec![SCHEDULED_TIME] },
-				PARA_ID.try_into().unwrap(),
-				NATIVE,
-				MultiLocation::new(1, X1(Parachain(PARA_ID.into()))).into(),
+				Box::new(destination.clone().into()),
+				Box::new(MultiLocation::default().into()),
+				Box::new(AssetPayment { asset_location: destination.into(), amount: 10 }),
 				call.clone(),
 				Weight::from_ref_time(100_000),
+				Weight::from_ref_time(200_000),
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:419]: Same question as above.

**Review by chrisli30 — CHANGES_REQUESTED** [review]
A lot to improve for this code change.

File: pallets/automation-time/src/tests.rs (hunk 12 above)
Comment (imstar15) [reply pallets/automation-time/src/tests.rs:359]: Because the schedule_xcmp_task_through_proxy function definition now includes the `overall_weight` parameter, I need to pass the parameter value here.

File: pallets/automation-time/src/tests.rs (hunk 13 above)
Comment (imstar15) [reply pallets/automation-time/src/tests.rs:405]: https://github.com/OAK-Foundation/OAK-blockchain/pull/366/files/e3f9a52c4885ed949e334faefb23e8d859b578ef..617511c19e9d300d7e55b676c4d01416f6355301#r1257652871

File: pallets/automation-time/src/tests.rs [hunk 14]
```diff
@@ -289,6 +293,33 @@ fn schedule_native_transfer_works() {
 	})
 }
 
+#[test]
+fn calculate_schedule_fee_amount_works() {
+	new_test_ext(START_BLOCK_TIME).execute_with(|| {
+		let destination = MultiLocation::new(1, X1(Parachain(PARA_ID)));
+		let delegator_account = AccountId32::new(DELEGATOR_ACCOUNT);
+		let execution_count = 2;
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/tests.rs:287]: How about the `times_of_task_execution`?

File: pallets/automation-time/src/tests.rs (hunk 11 above)
Comment (imstar15) [reply pallets/automation-time/src/tests.rs:303]: If the task is a xcmp task. the weight of the task is a const. 

It is defined here.

https://github.com/OAK-Foundation/OAK-blockchain/blob/69464509e9dd628485b6633df1b33718ce42cbab/pallets/automation-time/src/weights.rs#L311

It is not calculated based on the data of the task/action.

So, I think it should be fine to define the XCMP_TASK_REF_TIME constant as well.

File: pallets/automation-time/src/tests.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (chrisli30) [reply pallets/automation-time/src/tests.rs:287]: Let’s use `num_of_execution`, number of execution.

File: pallets/automation-time/src/tests.rs (hunk 12 above)
Comment (chrisli30) [reply pallets/automation-time/src/tests.rs:359]: okay, got it. That makes sense.

File: pallets/automation-time/src/fees.rs [hunk 15]
```diff
@@ -73,46 +73,50 @@ where
...
 			.ok_or(DispatchError::Token(BelowMinimum))?;
-		T::MultiCurrency::ensure_can_withdraw(self.currency_id.into(), &self.owner, fee)?;
+		T::MultiCurrency::ensure_can_withdraw(currency_id, &self.owner, fee)?;
 		Ok(())
 	}
 
 	/// Withdraw the fee.
 	fn withdraw_fee(&self) -> Result<(), DispatchError> {
-		let fee = self.execution_fee.saturating_add(self.xcmp_fee);
+		let fee = self.schedule_fee_amount.saturating_add(self.execution_fee_amount);
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/fees.rs:99]: We will modify the deduction logic in a later version, so don't worry about it in this PR.

File: pallets/automation-time/src/tests.rs [hunk 16]
```diff
@@ -31,12 +35,77 @@ use sp_runtime::{
...
+			encoded_call: vec![3, 4, 5],
+			encoded_call_weight: Weight::from_ref_time(100_000),
+			overall_weight: Weight::from_ref_time(200_000),
+			schedule_as: Some(delegator_account),
+			flow: InstructionSequence::PayThroughRemoteDerivativeAccount,
+		}
+	}
+}
+
+fn create_xcmp_task_action(options: XcmpTaskParams) -> ActionOf<Test> {
```
[the remark this answers is not in the document]
Comment (chrisli30) [reply pallets/automation-time/src/tests.rs:79]: I agree with Charles. I think this function is mostly used to construct the default parameters for tests so we don’t need to write a full list of parameters in every test. It’s not used to mock dev code.

File: pallets/automation-time/src/tests.rs [hunk 17]
```diff
@@ -280,29 +355,234 @@ fn schedule_transfer_with_dynamic_dispatch() {
...
-		let action: ActionOf<Test> = Action::XCMP {
-			destination,
-			schedule_fee: destination.into(),
-			execution_fee: AssetPayment { asset_location: destination.into(), amount: 100 },
-			encoded_call: vec![3, 4, 5],
-			encoded_call_weight: Weight::from_ref_time(100_000),
-			overall_weight: Weight::from_ref_time(200_000),
-			schedule_as: Some(delegator_account),
-			flow: InstructionSequence::PayThroughRemoteDerivativeAccount,
+		let num_of_execution = 2;
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:360]: This function is important. Could we use an arbitrary number such as 15, instead of 2? 2 is too common.

Comment (chrisli30) [reply pallets/automation-time/src/tests.rs:360]: Same applies to the below `num_of_execution` values. Maybe just use a random value from 1 to 20.

File: pallets/automation-time/src/tests.rs [hunk 18]
```diff
@@ -280,29 +355,234 @@ fn schedule_transfer_with_dynamic_dispatch() {
...
+
+		let action = create_xcmp_task_action(XcmpTaskParams {
+			schedule_fee: MultiLocation::new(
+				1,
+				X1(Parachain(<Test as Config>::SelfParaId::get().into())),
+			),
+			..XcmpTaskParams::default()
+		});
+		let fee_amount_1 = AutomationTime::calculate_schedule_fee_amount(&action, num_of_execution)
```
Comment (chrisli30) [inline pallets/automation-time/src/tests.rs:449]: Please use `fee_amount_abosolute` instead of `fee_amount_1`. The same goes to all `_1` and `_2`s. What are 1 and 2?

**Review by chrisli30 — CHANGES_REQUESTED** [review]
I have a couple of minor suggestions but apply to many tests.

File: pallets/automation-time/src/lib.rs [hunk 19]
```diff
@@ -394,20 +400,27 @@ pub mod pallet {
 			origin: OriginFor<T>,
 			provided_id: Vec<u8>,
 			schedule: ScheduleParam,
-			para_id: ParaId,
-			currency_id: T::CurrencyId,
-			xcm_asset_location: VersionedMultiLocation,
+			destination: Box<VersionedMultiLocation>,
```
[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/lib.rs:403]: Okay, I will update the comment

[the remark this answers is not in the document]
Comment (imstar15) [reply pallets/automation-time/src/lib.rs:406]: I have updated the comment. Please review.

**Review by chrisli30 — APPROVED** [review]
Looks good to me!
